MailChimp.Net/Logic/MemberLogic.cs

```diff
@@ -63,10 +65,71 @@ public MemberLogic(MailChimpOptions mailChimpConfiguration)
         ///     </paramref>
         /// includes an unsupported specifier. Supported format specifiers are listed in the Remarks section.
         /// </exception>
-        public async Task<Member> AddOrUpdateAsync(string listId, Member member)
+        public async Task<Member> AddOrUpdateAsync(string listId, Member member, IList<MarketingPermissionText> marketingPermissions = null)
         {
             using (var client = CreateMailClient($"{BaseUrl}/"))
             {
+                if (marketingPermissions != null)
+                {
+                    var getListResponse = await client.GetAsync($"{listId}").ConfigureAwait(false);
+                    await getListResponse.EnsureSuccessMailChimpAsync().ConfigureAwait(false);
+
+                    var list = await getListResponse.Content.ReadAsAsync<List>().ConfigureAwait(false);
+
+                    await getListResponse.EnsureSuccessMailChimpAsync().ConfigureAwait(false);
+
+                    if (list.MarketingPermissions)
+                    {
+                        var currentListMarketingPermissions = new List<MarketingPermission>();
+
+                        var members = (await GetResponseAsync(list.Id, null).ConfigureAwait(false))?.Members;
+
+                        if (!members.Any())
+                        {
+                            var dummyMember = new Member
+                            {
+                                EmailAddress = $"dummyMember{DateTime.Now.Ticks}@test.com",
+                                StatusIfNew = Status.Subscribed,
+                                Status = Status.Subscribed,
+                                MergeFields = new Dictionary<string, object>
+                            {
+                                { "FNAME", "DUMMY" },
+                                { "LNAME", "MEMBER" }
+                            }
+                            };
+
+                            var putDummyMemberToListResponse = await client.PutAsJsonAsync($"{listId}/members/{Hash(dummyMember.EmailAddress.ToLower())}", dummyMember).ConfigureAwait(false);
+
+                            await putDummyMemberToListResponse.EnsureSuccessMailChimpAsync().ConfigureAwait(false);
+
+                            var dummyMemberResponse = await putDummyMemberToListResponse.Content.ReadAsAsync<Member>().ConfigureAwait(false);
+
+                            currentListMarketingPermissions = dummyMemberResponse.MarketingPermissions.ToList();
+
+                            await DeleteAsync(list.Id, dummyMember.EmailAddress);
+                        }
+                        else
+                        {
+                            currentListMarketingPermissions = members.First().MarketingPermissions.ToList();
+                        }
+
+                        for (var i = 0; i < currentListMarketingPermissions.Count; i++)
```

Comment:
change this entire section to member.MarketingPermissions = currentListMarketingPermissions.Select(y=> {do work})